### ISSUE TITLE
Cannot Unmarshal cloudevents.Event in Background Function

### ISSUE THREAD

**Issue by rhodee** [cloudevents/sdk-go#306]
I have not been able to parse the `cloudevents.Event` type while using https://github.com/GoogleCloudPlatform/functions-framework-go to build my background function.

The `cloudevents.Event` throws an error (log below) when unmarshaling. The framework [Readme](https://github.com/GoogleCloudPlatform/functions-framework-go#enable-cloud-events) was used as a basis for the model below. 

My background function looks like this:

```go
package email

import (
	"context"
	"fmt"
	"os"

	cloudevents "github.com/cloudevents/sdk-go"
)

func Emailer(ctx context.Context, e cloudevents.Event)  {
	fmt.Fprintf(os.Stdout, "Sending email...\nData:%v\nContext:%v\n", e.Context, e.Data)
d}
```

The calling function:

```go
import (
	"fmt"
	"os"

	cloudevents "github.com/cloudevents/sdk-go"
	cecontext "github.com/cloudevents/sdk-go/pkg/cloudevents/context"
	"github.com/cloudevents/sdk-go/pkg/cloudevents/transport/pubsub"
)

func CallingTopic() {
	t, err := pubsub.New(ctx, pubsub.WithProjectID(projectID), pubsub.WithTopicID(topicID))
	ctx = cecontext.WithTopic(context.Background(), topicID)

	c, err := cloudevents.NewClient(t, cloudevents.WithTimeNow(), cloudevents.WithUUIDs())
	if err != nil {
		fmt.Fprintf(os.Stdout, "failed to create client, %s", err.Error())
		return
	}

	event := cloudevents.NewEvent(cloudevents.VersionV01) 

	event.SetType("com.example.com")
	event.SetSource("com.example/send/")
	_ = event.SetData(&payload{
		Test:  true,
	})

	_, _, err = c.Send(ctx, event)
}
```

Stackdriver Logging reports the following error:

Version `0.10.0`

```
"Error: unnknown spec version: \"\", while converting event data: {\"@type\":\"type.googleapis.com/google.pubsub.v1.PubsubMessage\",\"attributes\":{\"Content-Type\":\"application/cloudevents+json\"},\"data\":\"REMOVED"}"
```

Version `1.0.0`
```json
Error: unnknown spec version: "", while converting event data: {"@type":"type.googleapis.com/google.pubsub.v1.PubsubMessage","attributes":{"ce-datacontenttype":"application/json","ce-id":"f2afbe3c-16f2-4fa4-8521-2ca61193face","ce-source":"com.sendcoffee.svc.person","ce-specversion":"0.1","ce-time":"2020-02-15T08:27:04.459994603Z","ce-type":"com.sendcoffee.create-email"},"data":"REMOVED"}
```

I've tracked down the error log above to this line: https://github.com/cloudevents/sdk-go/blob/4db28ccdad5a61a483a22388d70d3fce729a1106/pkg/cloudevents/event_marshal.go#L73. I am confused however because I am setting the Version value as shown above.

I used the `cmd/samples/pubsub/sender` was the basis for my implementation above. Any attention you could provide would be appreciated. I can also open an issue in the other project as well if you think it is helpful.

Thank you.

Versions Tested

- [1.0.0](https://github.com/cloudevents/sdk-go/tree/v1.0.0)
- [0.10.0](https://github.com/cloudevents/sdk-go/tree/v0.10.0)

**Comment by n3wscott** [comment]
Pubsub support was added only after CloudEvents version 0.2 if I remember correctly. Try changing the following:

```go
event := cloudevents.NewEvent(cloudevents.VersionV01)
```

to 

```go
event := cloudevents.NewEvent(cloudevents.VersionV1) 
```

And you can continue using the goland cloudevents sdk 1.0.0 release.

I think the issue is in your cloudevents message: `"ce-specversion":"0.1"`, that is a very old version.

**Comment by rhodee** [comment]
I made the change as you suggested. I still get the same error.

```
"Error: unnknown spec version: \"\", while converting event data: 
{
  "@type": "type.googleapis.com/google.pubsub.v1.PubsubMessage",
  "attributes": {
    "Content-Type": "application/cloudevents+json"
  },
  "data": "REMOVED"
}
```

I am concerned that the framework project may not be using the latest stable version and perhaps this is the cause of the issue. Would you agree based on this line in go.mod? https://github.com/GoogleCloudPlatform/functions-framework-go/blob/master/go.mod#L7

If it helps I can raise an issue there as well.

**Comment by n3wscott** [comment]
That is not a valid cloudevent payload. I do not think gcp functions framework is producing cloudevents correctly. And the layers of processing are unclear to me. Is this via pubsub push? or some magic that the functions framework is doing for you? It might be that you don't need to parse the message and they don't want you to parse? 

I don't know.

**Comment by rhodee** [comment]
I am going to try the same request now that the functions framework has been updated and report my findings here thank you!

**Comment by slinkydeveloper** [comment]
Any updates on this? Can we close it?

**Comment by majelbstoat** [comment]
This is still a problem, though it's unclear if the issue is within the functions-framework, or in this sdk. This is the message emitted by the pub/sub local emulator after message creation using this SDK:

```json
{
  "subscription": "projects\/my-project\/subscriptions\/my-subscription",
  "message": {
    "data": "eyJ1c2VySWQiOiJBQUVCRlRpdXE0UFNEUEJ5UEEifQ==",
    "messageId": "1",
    "attributes": {
      "ce-source": "my-app",
      "ce-specversion": "1.0",
      "ce-type": "test.event.signIn",
      "ce-id": "MFkWEgrOHKs44AEADQ",
      "ce-time": "2020-05-24T18:38:18.7759848Z",
      "Content-Type": "application\/json"
    }
  }
}
```

Parsing fails with:

```
Error: specversion: unknown : ""
```

If this is a valid message, the problem would seem to be the functions-framework failing to unmarshal it. If it's invalid, it could be this SDK's message generation.

I opened [another issue](https://github.com/cloudevents/sdk-go/issues/518) which is somewhat related. (The http workaround I tried does not work because the message can't be parsed.)

I also opened an issue on [functions-framework-go](https://github.com/GoogleCloudPlatform/functions-framework-go/issues/26).

**Comment by majelbstoat** [comment]
the root cause for both could possibly be that Pub/Sub push doesn't send the cloudevent attributes as headers, but as attributes on the message.

https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/http/message.go#L50

m.format is nil because the content type is `application/json`, not `application/cloudevents+json`.
m.version is nil, because there's no `Ce-Specversion` header.

(I've now tested this both with real pub/sub and the emulator.)

**Comment by majelbstoat** [comment]
I ended up writing a custom format that appears to narrrowly do what I need, but I do feel like this should be supported in the SDK:

```go
package cloudevents

import (
	"encoding/json"
	"net/url"
	"time"

	"github.com/cloudevents/sdk-go/v2/event"
	"github.com/cloudevents/sdk-go/v2/types"
)

var JSON = jsonFmt{}

type PubSubPushEvent struct {
	Message struct {
		Data       []byte `json:"data"`
		Attributes struct {
			Source  string `json:"ce-source"`
			Version string `json:"ce-specversion"`
			Type    string `json:"ce-type"`
			ID      string `json:"ce-id"`
			Time    string `json:"ce-time"`
		} `json:"attributes"`
	} `json:"message"`
}

type jsonFmt struct{}

func (jsonFmt) MediaType() string { return event.ApplicationJSON }

func (jsonFmt) Marshal(e *event.Event) ([]byte, error) { return json.Marshal(e) }
func (jsonFmt) Unmarshal(b []byte, e *event.Event) error {
	pspe := &PubSubPushEvent{}
	err := json.Unmarshal(b, pspe)
	if err != nil {
		return err
	}

	t, err := time.Parse(time.RFC3339Nano, pspe.Message.Attributes.Time)
	if err != nil {
		return err
	}
	e.DataEncoded = pspe.Message.Data
	e.Context = &event.EventContextV1{
		ID: pspe.Message.Attributes.ID,
		Source: types.URIRef{URL: url.URL{
			Host: pspe.Message.Attributes.Source,
		}},
		Type: pspe.Message.Attributes.Type,
		Time: &types.Timestamp{Time: t},
	}
	return nil
}
```

**Comment by slinkydeveloper** [comment]
@majelbstoat since this issue you identified is an issue specific to pub/sub protocol and not this issue, can you create a new issue?

**Comment by majelbstoat** [comment]
Yep, I created this last week, before I saw this issue: https://github.com/cloudevents/sdk-go/issues/518

**Comment by slinkydeveloper** [comment]
Ok good i close this one, reopen if you need it
